### PR TITLE
[V1] [Hybrid] Support Minimax-Text-01 in V1 

### PR DESCRIPTION
**Pull request by tdoublep**
# Essential Elements of an Effective PR Description Checklist

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

## Purpose

This PR enables Minimax-Text-01 in vLLM V1. Currently only eager mode is supported. 

I've tried to keep the changes minimal. I hope this PR could serve as a reference/template to how to enable other non-mamba2 based hybrid models in V1. 

## Test Plan

Deploy using V0 from main:
```
vllm serve MiniMaxAI/MiniMax-Text-01 \
	--tensor-parallel-size 8 \
	--trust-remote-code \
	--quantization experts_int8  \
	--max_model_len 4096 \
	--dtype bfloat16 \
	--gpu-memory-utilization 0.95
```

Deploy using V1 from this PR:
```
VLLM_USE_V1=1 VLLM_ATTENTION_BACKEND=FLASHINFER vllm serve MiniMaxAI/MiniMax-Text-01 \
	--tensor-parallel-size 8 \
	--trust-remote-code \
	--quantization experts_int8  \
	--max_model_len 4096 \
	--dtype bfloat16 \
	--gpu-memory-utilization 0.95 \
	--no-enable-prefix-caching \
	--enforce-eager
```
Evaluation:
```
lm_eval   --model local-completions   \
	--model_args base_url=http://localhost:8000/v1/completions,tokenizer=MiniMaxAI/MiniMax-Text-01 \
	--tasks gsm8k  \
	--batch_size 128 \
	--num_fewshot 5 \
	--limit 500
```

## Test Result

V0:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.898|±  |0.0135|
|     |       |strict-match    |     5|exact_match|↑  |0.892|±  |0.0139|
```

V1:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.890|±  |0.0140|
|     |       |strict-match    |     5|exact_match|↑  |0.886|±  |0.0142|
```

### CODE REVIEW

File: vllm/model_executor/layers/lightning_attn.py [hunk 1]
```diff
@@ -477,7 +477,7 @@ def lightning_attention(q, k, v, ed, block_size=256, kv_history=None):
         ed = ed.view(1, -1, 1, 1)
 
     # Split the computation into chunks for better parallelism
-    m = 128 if d >= 128 else 64
+    m = 128 if d >= 128 else 32
```
Comment (tdoublep) [inline vllm/model_executor/layers/lightning_attn.py]: This change was needed to run this tiny model, need to revert it once tested on big model.

**Comment by qscqesze** [comment]
I'm very interested in the tiny model. Would you be able to share it?

**Comment by tdoublep** [comment]
@qscqesze sure, I'm using this one: https://huggingface.co/Goekdeniz-Guelmez/MiniMax01Text-Dev

**Comment by NickLucche** [comment]
I think you forgot to paste the gsm8k V1 results in the description, or I am just getting ahead of myself since it's still a draft, sorry in that case :)

**Comment by tdoublep** [comment]
@NickLucche yeah haha, I didn't paste them because they don't look good yet lol. Still debugging it

**Comment by tdoublep** [comment]
It seems to be specifically an issue when the blocks start to get recycled. I see good lm_eval results until blocks start running out and getting re-used. To me this suggests some issue with the way the state is being reset.

**Comment by tdoublep** [comment]
@NickLucche Fixed the problem and updated above :)

**Comment by tdoublep** [comment]
BTW the problem was related to trying to use fp32 for the linear attention state, but fp16 for the normal attention state. This is what happens on V0 right now, but didn't seem to work when I tried it in V1. Not sure if this is some inherent limitation, will think more on that.

**Comment by rogeryoungh** [comment]
I have tested your pull request, the accuracy on GSM8k is 0.920, and the average accuracy on MMLU is 0.846.

Deployment command:

```bash
python3 -m vllm.entrypoints.api_server --model /data/xxx/model/MiniMax-Text-01/ --tensor-parallel-size 8 --trust-remote-code --quantization experts_int8 --max_model_len 4096 --dtype bfloat16 --no-enable-prefix-caching --enforce-eager
```

I used the [sglang benchmark scripts](https://github.com/sgl-project/sglang/tree/main/benchmark). The detailed results are as follows:

GSM8k test:

```bash
python3 bench_other.py --num-questions 500 --num-shots 5 --backend vllm --port 8000 --host http://127.0.0.1
# ...
Accuracy: 0.920
Invalid: 0.000
Latency: 155.861 s
```

MMLU test:

```bash
python3 bench_other.py --nsub 200 --backend vllm --port 8000 --host http://127.0.0.1
# ...
Total latency: 1511.974
Average accuracy: 0.846
```

File: vllm/model_executor/models/minimax_text_01.py [hunk 2]
```diff
@@ -944,23 +1014,27 @@ def forward(self,
...
+                state_indices_tensor,
+            ) = self.minimax_cache.current_run_tensors(**kwargs)
+            if getattr(attn_metadata, "num_prefills", 0) > 0:
+                self._clear_prefill_cache(attn_metadata, minimax_cache_tensors,
+                                          **kwargs)
+
+            minimax_cache_params = MinimaxCacheParams(minimax_cache_tensors,
+                                                      state_indices_tensor)
+        else:
+            minimax_cache_params = None
```
Comment (NickLucche) [inline vllm/model_executor/models/minimax_text_01.py:1039]: qq: is this not needed in v1?

File: vllm/model_executor/models/minimax_text_01.py [hunk 3]
```diff
@@ -1321,3 +1405,29 @@ def load_basic_weight(name: str, loaded_weight: torch.Tensor,
 
             load_basic_weight(name, loaded_weight, self)
         return loaded_params
+
+    @classmethod
+    def get_mamba_state_shape_from_config(
+        cls,
+        vllm_config: "VllmConfig",
+        use_v1: bool = True,
+    ) -> tuple[tuple[int, int], tuple[int, int, int]]:
```
Comment (NickLucche) [inline vllm/model_executor/models/minimax_text_01.py]: type

**Review by NickLucche — APPROVED** [review]
lgtm, but let's wait for someone more familiar with this the model to chime-in. 
It's a bit of a pity we can pair this with a v0 deprecation task and just get rid of the v0 branching.

File: vllm/model_executor/models/minimax_text_01.py (hunk 2 above)
Comment (tdoublep) [reply vllm/model_executor/models/minimax_text_01.py:1039]: Nope

File: vllm/model_executor/models/minimax_text_01.py (hunk 3 above)
Comment (tdoublep) [reply vllm/model_executor/models/minimax_text_01.py]: Fixed

File: vllm/model_executor/models/minimax_text_01.py [hunk 4]
```diff
@@ -1321,3 +1379,29 @@ def load_basic_weight(name: str, loaded_weight: torch.Tensor,
...
+            - temporal_state_shape: Shape for state space model cache
+        """
+        parallel_config = vllm_config.parallel_config
+        hf_config = vllm_config.model_config.hf_config
+
+        state_shape = (hf_config.num_attention_heads //
+                       parallel_config.tensor_parallel_size,
+                       hf_config.head_dim, hf_config.head_dim)
+
+        return [state_shape]
```
[the remark this answers is not in the document]
Comment (tdoublep) [reply vllm/model_executor/models/minimax_text_01.py]: Fixed

**Comment by tdoublep** [comment]
@NickLucche Re: V0 code, I agree in general. There will still be a performance gap to V0 until we merge #21401, so I'm not sure we should remove V0 code until performance is at least matching. I think we may also want to wait until hybrid models are supported using FlashAttention (e.g., after #21549 is merged).

File: vllm/model_executor/models/minimax_text_01.py [hunk 5]
```diff
@@ -327,7 +330,16 @@ def jit_linear_forward_prefix(q: torch.Tensor,
...
 
 
-class MiniMaxText01LinearAttention(nn.Module):
+class MiniMaxText01LinearAttention(nn.Module, MambaBase):
+
+    @property
+    def mamba_type(self) -> str:
+        return "MiniMaxText01LinearAttention"
+
+    def get_state_shape(self) -> tuple[tuple[int, ...], tuple[int, ...]]:
```
Comment (Josephasafg) [inline vllm/model_executor/models/minimax_text_01.py:341]: Is there a way to use `MambaStateShapeCalculator` here?

Comment (tdoublep) [reply vllm/model_executor/models/minimax_text_01.py:341]: Yes there is. I will add that now the mamba1 PR is merged.

Comment (tdoublep) [reply vllm/model_executor/models/minimax_text_01.py:341]: Have done that now

**Comment by youkaichao** [comment]
cc @heheda12345

**Comment by qscqesze** [comment]
I think this code is fine. Great job! Looking forward to its early merge.

**Review by heheda12345 — APPROVED** [review]
LGTM! But can you move all ssm-related attention backends to `v1/attention/backends/mamba`? Either in this PR or a follow-up PR is OK for me. 

And let's do more refactor to make more integration more plugable in the future.

**Comment by heheda12345** [comment]
And thanks minimax team for verifying the correctness of this PR.

**Review by DarkLight1337 — COMMENTED** [review]
Thanks, can you open a follow-up PR to update the docs accordingly?

**Comment by zhiyuan1i** [comment]
> BTW the problem was related to trying to use fp32 for the linear attention state, but fp16 for the normal attention state. This is what happens on V0 right now, but didn't seem to work when I tried it in V1. Not sure if this is some inherent limitation, will think more on that.


Thank you for the elegant design and meticulous engineering in this work—your code has already become the go-to reference for many of us.

I’m currently aligning our in-house, multi-branch hybrid model for an upcoming sync back to main. We also maintain a dual-precision linear cache (some tensors stay in bf16 while others stay in fp32), so your implementation will be an invaluable blueprint for how to handle the mixed-precision paths cleanly. 


I believe the limitation exits  because:
`vllm/v1/kv_cache_interface.py`
```
@dataclass(frozen=True)
class MambaSpec(KVCacheSpec):
    shapes: tuple[tuple[int, ...], ...]
    dtype: torch.dtype
    page_size_padded: Optional[int] = None
    mamba_type: str = "mamba2"

    @property
    def page_size_bytes(self) -> int:
        num_elements = sum(prod(shape) for shape in self.shapes)
        page_size = num_elements * get_dtype_size(self.dtype)
        if self.page_size_padded is not None:
            assert self.page_size_padded >= page_size
            return self.page_size_padded
        return page_size

    def max_memory_usage_bytes(self, vllm_config: VllmConfig) -> int:
        # We allocate 1 block for each request now, so max_memory_usage_bytes is
        # the same as page_size_bytes.
        # Need to update this when supporting prefix caching.
        return self.page_size_bytes

```

This codes only consider one type?
